File: rb101_rb109/easy7/swap_case.rb
=begin
request: Write a method that takes a string as an argument and 
returns a new string in which every uppercase letter is replaced by its lowercase version,
 and every lowercase letter by its uppercase version. 
 All other characters should be unchanged.
rules: 
  - not use  String#swapcase
  - return a new string
data: array and string
algorithm:
  - call the .split method on the given string, to transform the string to array
  - call the .map method on this new array, passing in a block as argument
    - pass each element (small piece of string) of this array to the block
      -  call the .chars on each element (small piece of string), 
        - to return an array with each char
        - call .map method on each inner array
          - iterate through each char
            - if char is within alphabets
                - change lowercase to upcase 
                - change upcase to lowercase
            - if char is not alphabets
                - return the char unchaged
          - call .join method on the inner array, to transform the array back to string
          - assign the trasnsformed string to a local variable (swap_chars)
          - return the local variable (swap_chars) back to first map method
  - assign the new outer array to a local variable (swap_cases)
  - return swap_cases              
=end
def swapcase(str)
  swap_case = str.split.map do |inner_str|
    swap_chars = inner_str.chars.map do |char|   #["c", "b","a"]  => "cba"
      if /[a-z]/.match?(char)
        char.upcase
      elsif /[A-Z]/.match?(char)
        char.downcase
      else
        char
      end
    end
    swap_chars.join
  end
  swap_case.join(" ")
end

st = "CamEra"
p swapcase('CamelCase')
p swapcase('Tonight on XYZ-TV') 
p swapcase(st)
puts st

File: ls215/js210_string_processing/swap_case.js
/*
Question: a function that swap every letter with opposite case  - lower to upper; upper to lower
 arg: a string
 return a new string
algo: 
  - transform each char of the string
    - lowercase -> uppercase
    - uppercase -> lowercase
    - neither -> unchanged
  - return transfomred string
*/

/* first try
function swapCase(string) {
  return string.split('').map(char => {
    char = char.match(/[a-z]/) ? char.toUpperCase() : char.toLowerCase();
    return char;
  }).join('')
}
*/

function swapCase(string) {
  return string.split('').map(char => {
    if (/[a-z]/.test(char)) {
      return char.toUpperCase();
    } else if (/[A-Z]/.test(char)) {
      return char.toLowerCase();
    } else {
      return char;
    }
  }).join('');
}

console.log(swapCase('CamelCase'));
console.log(swapCase('Tonight on XYZ-TV'));